pkg/steps/template.go

```diff
@@ -421,18 +421,57 @@ func createOrRestartPod(podClient coreclientset.PodInterface, pod *coreapi.Pod)
 	return created, nil
 }
 
-func waitForPodDeletion(podClient coreclientset.PodInterface, name string, uid types.UID) error {
+func waitForTeardownToComplete(podClient coreclientset.PodInterface, name string, uid types.UID) error {
 	timeout := 300
 	for i := 0; i < timeout; i += 2 {
-		pod, err := podClient.Get(name, meta.GetOptions{})
-		if errors.IsNotFound(err) {
-			return nil
+		pod, err := checkPodExists(podClient, name, uid)
+		if err != nil {
+			return err
+		}
+		if !podTeardownCompleted(pod) {
+			log.Printf("Waiting for pod %s to complete teardown ... (%ds/%d)", pod.GetName(), i, timeout)
+			time.Sleep(2 * time.Second)
+		} else {
+			break
 		}
+	}
+	return nil
+}
+
+func checkPodExists(podClient coreclientset.PodInterface, name string, uid types.UID) (*coreapi.Pod, error) {
+	pod, err := podClient.Get(name, meta.GetOptions{})
+	if errors.IsNotFound(err) {
+		return nil, errors.NewGone("pod was removed")
+	}
+	if err != nil {
+		return nil, fmt.Errorf("could not retrieve deleting pod: %v", err)
+	}
+	if pod.UID != uid {
+		return nil, errors.NewBadRequest("mismatching UID")
+	}
+	return pod, nil
+}
+
+func waitForPodDeletion(podClient coreclientset.PodInterface, name string, uid types.UID) error {
+	timeout := 300
+	pod, err := checkPodExists(podClient, name, uid)
+	if err != nil {
+		return err
+	}
+
+	// Attempts to wait for teardown to complete
+	containerNames := podContainerNames(pod)
+	if sort.SearchStrings(containerNames, "teardown") < len(containerNames) {
+		err := waitForTeardownToComplete(podClient, name, uid)
 		if err != nil {
-			return fmt.Errorf("could not retrieve deleting pod: %v", err)
+			return err
 		}
-		if pod.UID != uid {
-			return nil
+	}
+
+	// Wait for pod to be completed
+	for i := 0; i < timeout; i += 2 {
+		if _, err := checkPodExists(podClient, name, uid); err != nil {
```

Comment:
This is getting tricky, now we never end successfully.

Reply:
It would throw `errors.NewGone("pod was removed")` and immediately return if the pods is gone, isn't it?

Reply:
Well yeah, but previously ` waitForPodDeletion` returned `nil` when the pod was "successfully" not found, so the callsites probably rely on that.

Reply:
Ah, agree, `abb16d4` fixes that - `waitForPodDeletion` now returns nil in this case